File: C#/SubmodularFunction/Submodular/SFMResult.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Onigiri.Submodular
{

    public class SFMResult
    {

        const int OutputLength = 20;
        SubmodularOracle oracle;
        //Stopwatch timer;
        //long memoOracleCallTime; //for calculation
        //long memoBaseCallTime;  //for calculation
        //long memoReductionTime; //for calculation
        double[] x;

        string[] labels = new string[]{"N", 
            "reduced N", 
            "Iteration", 
            "Execurtion Time", 
            "Oracle Time",
            "Reduction Time", 
            "Oracle Call", 
            "Base Call", 
            "Reduction Call", 
            "Minimum Value", 
            "Minimizer", 
            "Dual Value", 
            };


        internal SFMResult(SubmodularOracle oracle)
        {
            this.oracle = oracle;
            //this.N = n;
            //timer = new Stopwatch();
            //timer.Start();
        }

        private void WriteLable(string path)
        {
            if (File.Exists(path))
            {
                return;
            }
            var sw = new StreamWriter(path,false);
            for (int i = 0; i < labels.Length; i++)
            {
                if (i != 0)
                {
                    sw.Write('\t');
                }
                sw.Write(labels[i]);
            }
            sw.WriteLine();
            sw.Close();
        }

        public void Output(string path,bool withX = true)
        {
            StreamWriter streamWriter;
            WriteLable(path);
            while (true)
            {
                try { streamWriter = new StreamWriter(path,true); }
                catch
                {
                    continue;
                }
                break;
            }//
            OutputVariable("N", BaseN, streamWriter);
            OutputVariable("reduced N", N, streamWriter);
            OutputVariable("Iteration", Iteration, streamWriter);
            OutputVariable("Execurtion Time", ExecutionTime, streamWriter);
            OutputVariable("Oracle Time", OracleTime, streamWriter);
            OutputVariable("Reduction Time", ReductionTime, streamWriter);
            OutputVariable("Oracle Call", OracleCall, streamWriter);
            OutputVariable("Base Call", BaseCall, streamWriter);
            OutputVariable("Reduction Call", ReductionTime, streamWriter);
            OutputVariable("Minimum Value", MinimumValue, streamWriter);
            OutputVariable("Minimizer", Minimizer, streamWriter);
            OutputVariable("Dual Value", DualValue, streamWriter,false);
            if (withX)
            {
                OutputArray("x", X, streamWriter);                
            }//if
            streamWriter.Close();
        }

        private void OutputArray<T>(string name, T[] array, StreamWriter streamWriter)
        {
            streamWriter.WriteLine(name.PadRight(OutputLength) + " : " );
            if (array==null)
            {
                return;
            }//if
            foreach (var element in array)
            {
                streamWriter.WriteLine(element.ToString());
            }//foreach element
        }

        private void OutputVariable<T>(string name, T value,StreamWriter streamWriter,bool flg = true)
        {
            //streamWriter.Write(name.PadRight(OutputLength) + " : " + value.ToString());
            streamWriter.Write(value.ToString());
            if (flg)
            {
                streamWriter.Write('\t');
            }
            else
            {
                streamWriter.WriteLine();
            }
        }

        internal void SetResult(double[]x,string minimizer,double minimumValue,long iteration,long executionTime,
            ConvexComponents component,SubmodularOracle oracle)
        {
            //timer.Stop();
            //ExecutionTime = timer.ElapsedMilliseconds;
            this.x = (x == null ? null: (double[])x.Clone());
            this.Minimizer = minimizer;
            this.MinimumValue = minimumValue;
            this.DualValue = (x == null ? 0 : x.Where(val => val <= 0).Sum());
            this.Iteration = iteration;
            this.ExecutionTime = executionTime;

            this.OracleCall = oracle.OracleCall;
            this.BaseCall = oracle.BaseCall;
            this.OracleTime = oracle.OracleTime;
        }
        
        //internal void StartOracle()
        //{
        //    memoOracleCallTime = timer.ElapsedMilliseconds;
        //    OracleCall++;
        //}

        //internal void StopOracle()
        //{
        //    OracleTime += timer.ElapsedMilliseconds - memoOracleCallTime;
        //}

        //internal void StartBase()
        //{
        //    memoBaseCallTime = timer.ElapsedMilliseconds;
        //    BaseCall++;
        //}

        //internal void StopBase()
        //{
        //    OracleTime += timer.ElapsedMilliseconds - memoBaseCallTime;
        //}

        //internal void StartReduction()
        //{
        //    memoReductionTime = timer.ElapsedMilliseconds;
        //    ReductionCall++;
        //}

        //internal void StopReduction()
        //{
        //    ReductionTime += timer.ElapsedMilliseconds - memoReductionTime;
        //}

        public string Minimizer
        {
            get;
            private set;
        }

        public double MinimumValue
        {
            get;
            private set;
        }

        public double[] X
        {
            get { return (x == null ? null : (double[])x.Clone()); }
        }

        public long Iteration
        {
            get;
            private set;
        }

        public long OracleCall
        {
            get;
            private set;
        }

        public long BaseCall
        {
            get;
            private set;
        }

        public long ReductionCall
        {
            get;
            private set;
        }

        public long ExecutionTime
        {
            get;
            set;
        }

        public long ReductionTime
        {
            get;
            set;
        }

        public long OracleTime
        {
            get;
            set;
        }

        public double DualValue
        {
            get;
            set;
        }

        public int BaseN
        {
            get { return oracle.BaseN; }
        }

        public int N
        {
            get { return oracle.N; }
        }


    }
}

File: C++/SubmodularFunction/SubmodularFunction/SFMResult.cpp
#include "Submodular.h"

using namespace OnigiriSubmodular;

SFMResult::SFMResult(int n){
	outputLength = 20;
	SFMResult::n = n;
	iteration=0;
	oracleCall=0;
	baseCall=0;
	reductionCall=0;
	executionTime=0;
	reductionTime=0;
	oracleTime=0;
	startTime = clock();
	x = NULL;
}

SFMResult::SFMResult(SFMResult &result){
	SFMResult::outputLength = result.outputLength;
	SFMResult::minimizer = result.minimizer;
	SFMResult::	minimumValue= result.minimumValue;
	SFMResult:: iteration= result.iteration;
	SFMResult::oracleCall=result.oracleCall;
	SFMResult:: baseCall = result.baseCall;
	SFMResult:: reductionCall = result.reductionCall;
	SFMResult:: executionTime= result.executionTime;
	SFMResult:: reductionTime=result.reductionTime;
	SFMResult::oracleTime= result.oracleTime;
	SFMResult:: dualValue = result.dualValue;
	SFMResult:: n = result.n;
	SFMResult:: x = new double[n];
	result.X(SFMResult::x);
}

SFMResult::	~SFMResult(){
	if(x!=NULL){
		delete[] x;
	}
}

void SFMResult:: SetResult(double* x,string &minimizer,double minimumValue,long iteration,SubmodularOracle &oracle){
	executionTime = clock()-startTime;
	SFMResult:: minimizer =minimizer;
	SFMResult::minimumValue = minimumValue;
	SFMResult::iteration = iteration;
	SFMResult::dualValue = 0;
	SFMResult::oracleCall = oracle.oracleCall;
	SFMResult::baseCall = oracle.baseCall;
	SFMResult::oracleTime= oracle.oracleTime;
	if(x!=NULL){
		SFMResult::x = new double[n];
		for(int i=0;i<n;i++){
			SFMResult::x[i] = x[i];
			if(x[i]<=0){
				SFMResult::dualValue +=x[i];
			}
		}
	}else{
		SFMResult::x = NULL;
	}
}

void SFMResult::OutputArray(string tittle, double* x,ofstream& file){
	tittle.resize(SFMResult::outputLength,' ');
	file<<tittle<<" : "<<endl;
	if(x!=NULL){
		for(int i=0;i<n;i++){
			file<<Converter::DoubleToString(x[i])<<endl;
		}
	}
}

void SFMResult::OutputVariable(string tittle, string value,ofstream& file){
	tittle.resize(SFMResult::outputLength,' ');
	file<<tittle<<" : "<<value<<endl;
}

void SFMResult::OutputVariable(string tittle, long value,ofstream& file){
	tittle.resize(SFMResult::outputLength,' ');
	file<<tittle<<" : "<<Converter::IntToString(value)<<endl;
}

void SFMResult::OutputVariableDouble(string tittle, double value,ofstream& file){
	tittle.resize(SFMResult::outputLength,' ');
	file<<tittle<<" : "<<Converter::DoubleToString(value)<<endl;
}

void SFMResult::Output(string path,bool withX){	
	ofstream file(path,ios_base::app);
	OutputVariable("N",N(),file);
	OutputVariable("Iteration", Iteration(), file);
	OutputVariable("Execurtion Time", ExecutionTime(), file);
	OutputVariable("Oracle Time", OracleTime(), file);
	OutputVariable("Reduction Time", ReductionTime(), file);
	OutputVariable("Oracle Call", OracleCall(), file);
	OutputVariable("Base Call", BaseCall(), file);
	OutputVariable("Reduce Call",ReductionCall(),file);
	OutputVariableDouble("Minimum Value", MinimumValue(), file);
	OutputVariable("Minimizer", Minimizer(), file);
	if(withX){
		OutputVariableDouble("Dual Value", DualValue(), file);
	}
	double* x = new double[N()];
	X(x);
	OutputArray("x", x, file);
	delete[] x;
	file.close();
}

//void SFMResult::  StartOracle(){
//
//	memoOracleCallTime = clock();
//	oracleCall++;
//}
//
//void SFMResult::  StopOracle(){
//	oracleTime += clock() - memoOracleCallTime;
//}
//
//void SFMResult::  StartBase(){
//	memoBaseCallTime = clock();
//	baseCall++;
//}
//
//void SFMResult::  StopBase(){
//	oracleTime += clock() - memoBaseCallTime;
//}
//
//void SFMResult::  StartReduction(){
//	memoReductionTime = clock();
//	reductionCall++;
//}
//
//void SFMResult::  StopReduction(){
//	reductionTime += clock() - memoReductionTime;
//}

string SFMResult:: Minimizer(){
	return minimizer;
}

double SFMResult::  MinimumValue(){
	return minimumValue;
}

void SFMResult::  X(double* &x){
	if(SFMResult::x==NULL){
		delete[]x;
		x = NULL;
	}else{
		for(int i=0;i<n;i++){
			x[i] = SFMResult::x[i];
		}
	}
}

long SFMResult::  Iteration(){
	return iteration;
}

long SFMResult::  OracleCall(){
	return oracleCall;
}

long SFMResult::  BaseCall(){
	return baseCall;
}

long SFMResult::  ReductionCall(){
	return reductionCall;
}

long SFMResult::  ExecutionTime(){
	return executionTime;
}

long SFMResult::  ReductionTime(){
	return reductionCall;
}

long SFMResult::  OracleTime(){
	return oracleTime;
}

double SFMResult:: DualValue(){
	return dualValue;
}

int SFMResult::  N(){
	return n;
}
